Visualizing repositories using Pygal (#4121)
section-engineering-education/engineering-education: content/articles/visualizing-repositories-using-pygal/index.md

---
layout: engineering-education
status: publish
published: true
url: /visualizing-repositories-using-pygal/
title: Visualizing repositories using Pygal
description: In this article, you'll learn how to construct a visualization that shows the comparative popularity of Python works on GitHub using Pygal.
author: duncan-ndegwa
date: 2021-09-03T00:00:00-04:35
topics: [API]
excerpt_separator: <!--more-->
images:

 - url: engineering-education\content\articles\visualizing-repositories-using-pygal\hero.jpg
   alt: Visualizing repositories using Pygal hero image
---

### Introduction
Pygal allows us to create a range of graphs and charts. In this article, you'll learn how to construct a visualization that shows the comparative Python's prominence works on github.com. We'll create an interactive bar chart, where the height of each bar will represent the sum of stars earned by the project. By clicking a bar, you'll be sent to the project's GitHub page.
<!--more-->

### Prerequisites

> You'll need some Python skills to follow along with this tutorial. You must be in a position to work with web APIs in Python. Also, you need to have [PIP](https://pip.pypa.io/en/stable/installation/) installed.

Reviewer: ```suggestion
You'll need some Python skills to follow along with this tutorial. You must be in a position to work with web APIs in Python. Also, you need to have [PIP](https://pip.pypa.io/en/stable/installation/) installed.
```
You can break the prerequisites as a list items:
Some knowledge of Python
Knowledge on Web APIs
Pip installed

### Installing pygal
To install Pygal, run the command below;

```python
pip install pygal
```

You can visit [Pygal's official site](http://www.pygal.org/en/stable/installing.html) for more information on the installation of Pygal.

### The pygal gallery
The Pygal gallery is a collection of charts that Pygal can generate.

> Check the chart types page to explore what types of charts Pygal can generate: Select **Documentation**, finally **Chart types** in [pygal](http://www.pygal.org/). You can examine how the visualizations are created by looking at the source code for each case.

Reviewer: ```suggestion
Check the chart types page to explore what types of charts Pygal can generate: Select **Documentation**, finally **Chart types** in [pygal](http://www.pygal.org/). You can examine how the visualizations are created by looking at the source code for each case.
```
You can also reorder this statement as this ....
"To grasp more of the chart types that can be generated by Pygal, head over to the documentation tab on the chart type page. (issue the link).
Additionally, you can examine how the visualizations are created by exploring the source code for each.

To get this done, you can run the code below:

```python
import requests
import pygal
from pygal.style import LightColorizedStyle as LCSe, LightenStyle as LSe
#  Start an API request.
url = 'https://api.github.com/search/repositories?q=language:python&sort=star'
request = requests.get(url)
print("Status  code:", request.status_code)
# In a variable, save the API response.
response_dictionary = request.json()
print("Total-repos:", response_dictionary['total_count'])
# Investigate the repository's contents.
repos_dicts = response_dictionary['items']
names, stars = [], []
for repos_dict in repos_dicts:
    names.append(repos_dict['name'])
    stars.append(repos_dict['stargazers_count'])
# Construct a visual picture
our_style = LSe('#333366', base_style=LCSe)
chart = pygal.Bar(style=our_style, x_label_rotation=45, show_legend=False)
chart.title = 'Python Works with the Most Stars on GitHub'
chart.x_labels = names
chart.add('', stars)
chart.render_to_file('py_repos.svg')
```

> To begin, we `import pygal` as well as the `Pygal styles` we'll require for the chart. We'll keep printing the status of the API request return and the total amount of repositories detected to see if the API call was successful. 

Reviewer: ```suggestion
To begin, we `import pygal` as well as the `Pygal styles` that we'll require for the chart. We'll keep printing the status of the API request return and the total amount of repositories detected to see if the API call was successful. 
```

> First, we generate 2 blank lists to house the data that will be used in the chart. We’ll use the name, and the number of stars of each project to label and determine the height of the bars respectively. The name of each work, as well as the amount of stars it has, are added to these lists inside the loop.

Reviewer: ```suggestion
First, we generate two blank lists to house the data that will be used in the chart. We’ll use the name, and the number of stars of each project to label and determine the height of the bars respectively. The name of each work, as well as the amount of stars it has, are added to these lists inside the loop.
```

> Next, we create a style based on a blue colored tint using the `LightenStyle` class `(alias LSe)`.In order to use the `LightColorizedStyle class`(`alias LCSe`), we additionally give the base style parameter. After that, we create a simple chart with Bar() and provide it our_style.We further give two extra style arguments: we rotate the labels 45 degrees along the x-axis (`x label rotation=45 `) and conceal the legend since we're just showing one series on the graph (`show legend=False`). The `x labels` property is assigned to the list names, and the graph is given a title. 

Reviewer: This section is a bit difficult to read though, maybe for beginners, you can use some  connectors to break complexity.

> We give an empty string for the label once we load the data since we don't require this data series to be tagged. Below is the graph that resulted:

Reviewer: Please avoid the 'we' clause. you have used it a couple of times.

myconfig.major_label_font_size = 17 #set the text size for the major label(The major labels are just the labels on the y-axis)
myconfig.truncate_label = 15  #To reduce the length of long project names to 15 characters, employ truncate label.
myconfig.show_y_guides = False  #set show_y_guides to False hide the horizontal lines on the graph
myconfig.width = 1000   #Choose a custom width for the graph so that it takes up maximum browser area
chart = pygal.Bar(myconfig, style=our_style)  #make an instance of Bar, and pass myconfig and style as arguments, respectively
chart.title = 'Python Projects with the Most Stars on GitHub'
chart.x_labels = names
chart.add('', stars)
chart.render_to_file('py_repos.svg')
```

The restyled chart is shown below:

![The styling for the chart has been refined.](/engineering-education/visualizing-repositories-using-pygal/refining.jpeg)

### Customizing Tooltips
Floating a mouse above a single bar displays data it represents. This is known as a tooltip, and it usually displays the tally of stars that a project possesses. Let's make a custom tooltip to display the descriptions of each project.

> Let's have a look at a simple example that uses the first three projects plotted singly with custom labels for each bar. Rather than passing a list of values to `add()`, we'll pass a list of dictionaries with the code below:

Reviewer: ```suggestion
Let's have a look at an example that uses the first three projects plotted singly with custom labels for each bar. Rather than passing a list of values to `add()`, we'll pass a list of dictionaries with the code below:
```

```python
import pygal
from pygal.style import LightColorizedStyle as LCSe, LightenStyle as LSe
our_style = LSe('#333366', base_style=LCSe)
chart = pygal.Bar(style=our_style, x_label_rotation=45, show_legend=False)
chart.title = 'Projects in Python '
chart.x_labels = ['public-apis', 'system-design-primer', 'Python']
plot_dictoinaries = [
 {'value': 144904, 'label': 'Description of public-apis.'},
 {'value': 139818, 'label': 'Description of system-design-primer.'},
 {'value': 113616, 'label': 'Description of Python.'},
 ]
chart.add('', plot_dictoinaries)
chart.render_to_file('bar_desc.svg')
```

> We create a `plot_dictoinaries` list with three dictionaries: one for the `public_apis` work, one for the `system_design_primer` work, and one for `Python`. There are two keys in every dictionary: `value` and `label`. Pygal calculates the height of each bar using the number linked with `value`, and it creates the **tooltip** for each bar using the string connected with `label`. The first dictionary, for instance, will generate a bar indicating a project with `144904` stars, with the tooltip `"Description of public-apis."`

Reviewer: ```suggestion
We create a `plot_dictionaries` list with three dictionaries: one for the `public_apis` work, the other for  `system_design_primer` work, and the last one  for `Python`. There are two keys in every dictionary: `value` and `label`. Pygal calculates the height of each bar using the number linked with `value`, and it creates the **tooltip** for each bar using the string connected with `label`. The first dictionary, for instance, will generate a bar indicating a project with `144904` stars, with the tooltip `"Description of public-apis."`
```
